### ISSUE TITLE
Implementing quality indicators in DESDEO 2.0

### ISSUE THREAD

**Issue by gialmisi** [industrial-optimization-group/DESDEO#172]
# Update 12.6.2025:

As our research group has various and urgent needs for different indicators, a priority list of indicators is given below with the most urgent indicator to be implemented on top. The state of each indicator (waiting to be implemented 🔴, being implemented 🟡, or done 🟢), its name, a reference for the indicator, and possible additional information, are given below.

If you start implementing any indicator, please update its status in the list below and assign yourself to the issue. Leaving a comment notifying other of you intentions is also a good idea.

## Indicator list

- [x] 🟢 R-metric, [reference](https://ieeexplore.ieee.org/abstract/document/8049301), the indicator has different variants: R-HV , which uses Hypervolume, and R-IGD which uses Inverted Generational Distance
- [ ] 🟡 PHI, [reference](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=10115283), the indicator can measure the learning and the decision phases of an interactive method separately, it is based on the Hypervolume
- [x]  🟢 IGD+, [reference](https://www.sciencedirect.com/science/article/pii/S0377221720309620?via%3Dihub), the indicator is also known as the modified IGD
- [x] 🟢 R2 metric, [reference](https://link.springer.com/chapter/10.1007/978-3-540-88908-3_14), described in Section 14.4.8 in the reference
- [ ] 🟡 Epsilon indicator, [reference](https://doi.org/10.1007/978-3-540-30217-9_84), old implementation in [desdeo-tools](https://github.com/industrial-optimization-group/desdeo-tools/blob/master/desdeo_tools/utilities/quality_indicator.py)

### Template for adding indicators

Use this template to add new items to the above list:

```
- [ ] 🔴🟡🟢, INDICATOR NAME, [reference](link-to-reference), OPTIONAL ADDITIONAL INFO
```

---

We are currently missing quality indicators in version 2.0 of DESDEO. At least a couple indicators need to be implemented, documented, and tested.

A paper describing what quality indicators are in the context of (interactive) multiobjective optimization can be found [here](https://jyx.jyu.fi/handle/123456789/95813)

How indicators have been implemented in the older version of DESDEO, can be found here: https://github.com/industrial-optimization-group/desdeo-tools/blob/master/desdeo_tools/utilities/quality_indicator.py

It would be desirable to have at least the quality indicator described in the linked paper above, and the hypervolume and epsilon indicators. These, and other indicators, are described in this [paper](https://ieeexplore.ieee.org/document/1197687) (if you do not have access to the paper, please send a DM).

Most likely, quality indicators will be implemented in `desdeo-tools`. Before a potential contributor starts working on this issue, it is desirable that they shortly describe their plan of action in a reply to this issue, and tagging @gialmisi and @light-weaver in the post.

**Comment by light-weaver** [comment]
So I started added some indicators in this commit: https://github.com/industrial-optimization-group/DESDEO/commit/b6594edccc10699cb3ac230c7c5eeec1d481b22d, followed by fixes and tests.

The older DESDEO framework used the walking fish group's implementation of hypervolume (hvwfg on PyPI). That package is now terribly out of date and poetry refuses to install it. Moreover, poetry refuses to build it from source too due to similar reasons. The uv package manager can build and install it from github, but I didn't wanna replace the whole toolchain to have this single indicator.

I ended up choosing pymoo for their HV implementation. I think they get theirs from DEAP. 

I also implemented some distance based indicators like GD, IGD, and averaged hausdorff distance. 

If you see any issues with the choices I made, feel free to change things. Otherwise, just keep adding more indicators. Make sure to use similar assumptions (about normalization, etc) across all implementations.

Also, now that we have pymoo itself as a dependency, I may add versions of IOPIS that use their no-preference EAs and make them interactive.

**Comment by light-weaver** [comment]
As I mentioned, I used pymoo for their HV implementation. It has led to some currently unresolvable issues with my virtual environment. Moreover, [it seems that their HV implementation is. quite slow at the momemt compared to moocore](https://github.com/anyoptimization/pymoo/issues/667). I did not want to try using [moocore](https://github.com/multi-objective/moocore/tree/main/python) myself yet as it is quite new (and is missing installation instructions afaik). The maintainers of moocore are quite experienced with creating good software, so I am sure these minor issues will be resolved soon. If compiling/installing moocore isn't an issue, use that for HV and IGD+. 

Eventually, we may also want to benchmark the desdeo implementation of dominance checks against moocore's. It may be worth it to use theirs. But I'll come back to it when I have time.

**Comment by light-weaver** [comment]
@gialmisi Did someone else from the team ask for that specific R metric implementation? I meant the R2 metric ([original work](https://www.imm.dtu.dk/documents/ftp/tr98/tr07_98.abstract.html), [easier read](https://link.springer.com/chapter/10.1007/978-3-540-88908-3_14) (section 14.4.8)).

**Comment by gialmisi** [comment]
> [@gialmisi](https://github.com/gialmisi) Did someone else from the team ask for that specific R metric implementation? I meant the R2 metric ([original work](https://www.imm.dtu.dk/documents/ftp/tr98/tr07_98.abstract.html), [easier read](https://link.springer.com/chapter/10.1007/978-3-540-88908-3_14) (section 14.4.8)).

Yes, @giomara-larraga did. Is the source given by you (@light-weaver) equivalent?

**Comment by giomara-larraga** [comment]
> > [@gialmisi](https://github.com/gialmisi) Did someone else from the team ask for that specific R metric implementation? I meant the R2 metric ([original work](https://www.imm.dtu.dk/documents/ftp/tr98/tr07_98.abstract.html), [easier read](https://link.springer.com/chapter/10.1007/978-3-540-88908-3_14) (section 14.4.8)).
> 
> Yes, [@giomara-larraga](https://github.com/giomara-larraga) did. Is the source given by you ([@light-weaver](https://github.com/light-weaver)) equivalent?

It is not the same metric, if @light-weaver needs R2, it should be added to the list

**Comment by gialmisi** [comment]
Communication issue. Everything is fine. I added the R2 metric to the list.

**Comment by MLopez-Ibanez** [comment]
> As I mentioned, I used pymoo for their HV implementation. It has led to some currently unresolvable issues with my virtual environment. Moreover, [it seems that their HV implementation is. quite slow at the momemt compared to moocore](https://github.com/anyoptimization/pymoo/issues/667). I did not want to try using [moocore](https://github.com/multi-objective/moocore/tree/main/python) myself yet as it is quite new (and is missing installation instructions afaik). The maintainers of moocore are quite experienced with creating good software, so I am sure these minor issues will be resolved soon. If compiling/installing moocore isn't an issue, use that for HV and IGD+.
> 

Hi, I added some basic installation instructions: https://github.com/multi-objective/moocore/tree/main/python#install

If anything is missing or not clear, please open an issue. 

The `moocore` package itself is quite new but the codes are not. It wraps C codes that I have been using for a decade or more, it is also based on code from the `eaf` R package (which is already obsolete by `moocore`) and it has an extensive testsuite for the C code: https://github.com/multi-objective/testsuite and Python and R have their own testsuites.

The reason why I have not set the version to 1.0 is that I would like to get feedback from downstream users (pymoo, DEAP, BoTorch, various R packages, and hopefully DESDEO) of what the API should look like. For example, I added an object oriented hypervolume API (https://multi-objective.github.io/moocore/python/reference/generated/moocore.Hypervolume.html) because that seems popular in Python packages. So if there is something in the current API or docs that is inconvenient or could be improved, please let me know or, even better, send a pull request. 

Currently there is no R2-indicator in `moocore` simply because I have never had the need to use it.  If somebody contributes one in Python, I'll be happy to merge it and add you to the list of contributors. And eventually, I will get around to re-implement it in C for speed so that it is also available in R (and eventually other languages). 

`moocore` is not a project from a single group or a single university. There is no paper yet and perhaps never will (if there is, I'd like to follow scikit-learn and add all contributors to version 1.0 as co-authors).

Even with its current non-state-of-the-art algorithms (and implementing the state-of-the-art ones is a goal of the project), `moocore` is order of magnitudes faster than most python packages (I only know one package that is comparable and it is because they have an underlying C++ implementation), not only in hypervolume but in all indicators. Please compare your IGD+ implementation with the one in `moocore`: https://multi-objective.github.io/moocore/python/reference/functions.metrics.html#inverted-generational-distance-igd-and-igd-and-averaged-hausdorff-distance

You can use the datasets from the testsuite to do so.

Think about all the accumulated computation cycles that are being wasted because of people re-implementing their own basic implementations (sometimes buggy ones). Would you advocate re-implementing your own array library instead of using Numpy or your own basic ML library instead of using scikit-learn? 

The goal of `moocore` is not to replace pymoo, DESDEO or any of those libraries, but simply to provide high-quality basic building blocks to make building such libraries easier.

**Comment by maiemile** [comment]
Could the epsilon indicator be added to the indicator list, it would be useful for IBEA. As already mentioned above, it has already been implemented in a previous version of DESDEO [here](https://github.com/industrial-optimization-group/desdeo-tools/blob/master/desdeo_tools/utilities/quality_indicator.py). The mathematical formula is given [here](https://doi.org/10.1007/978-3-540-30217-9_84) for example.
